esp-hal/src/dma/gdma.rs

```diff
@@ -606,3 +613,151 @@ impl<'d> Dma<'d> {
+            }
+        }
+
+        /// Create a new Mem2Mem instance.
+        ///
+        /// # Safety
+        ///
+        /// You must insure that your not using DMA for the same peripheral and
+        /// that your the only one using the DmaPeripheral.
+        pub unsafe fn new_unsafe(
+            mut channel: Channel<'d, C, MODE>,
+            peripheral: DmaPeripheral,
+            tx_descriptors: &'static mut [DmaDescriptor],
+            rx_descriptors: &'static mut [DmaDescriptor],
+        ) -> Self {
+            channel.tx.init_channel();
+            channel.rx.init_channel();
+            Mem2Mem {
+                channel,
+                peripheral,
+                tx_chain: DescriptorChain::new(tx_descriptors),
+                rx_chain: DescriptorChain::new(rx_descriptors),
+            }
+        }
+
+        /// Start a memory to memory transfer.
+        pub fn start_transfer<'t, TXBUF, RXBUF>(
+            &mut self,
+            tx_buffer: &'t TXBUF,
+            rx_buffer: &'t mut RXBUF,
+        ) -> Result<DmaTransferRx<Self>, DmaError>
+        where
+            TXBUF: ReadBuffer<Word = u8>,
+            RXBUF: WriteBuffer<Word = u8>,
+        {
+            let (tx_ptr, tx_len) = unsafe { tx_buffer.read_buffer() };
+            let (rx_ptr, rx_len) = unsafe { rx_buffer.write_buffer() };
+            self.tx_chain.fill_for_tx(false, tx_ptr, tx_len)?;
+            self.rx_chain.fill_for_rx(false, rx_ptr, rx_len)?;
+            unsafe {
+                self.channel
+                    .tx
+                    .prepare_transfer_without_start(self.peripheral, &self.tx_chain)?;
+                self.channel
+                    .rx
+                    .prepare_transfer_without_start(self.peripheral, &self.rx_chain)?;
+                self.channel.rx.set_mem2mem_mode(true);
+            }
+            self.channel.tx.start_transfer()?;
+            self.channel.rx.start_transfer()?;
+            Ok(DmaTransferRx::new(self))
+        }
+    }
+
+    impl<'d, C, MODE> Drop for Mem2Mem<'d, C, MODE>
+    where
+        C: ChannelTypes,
+        MODE: crate::Mode,
+    {
+        fn drop(&mut self) {
+            self.channel.rx.set_mem2mem_mode(false);
+        }
+    }
+
+    impl<'d, C, MODE> DmaSupport for Mem2Mem<'d, C, MODE>
+    where
+        C: ChannelTypes,
+        MODE: crate::Mode,
+    {
+        fn peripheral_wait_dma(&mut self, _is_tx: bool, _is_rx: bool) {
+            while !self.channel.rx.is_done() {}
+        }
+
+        fn peripheral_dma_stop(&mut self) {
+            unreachable!("unsupported")
+        }
+    }
+
+    impl<'d, C, MODE> DmaSupportRx for Mem2Mem<'d, C, MODE>
+    where
+        C: ChannelTypes,
+        MODE: crate::Mode,
+    {
+        type RX = C::Rx<'d>;
+
+        fn rx(&mut self) -> &mut Self::RX {
+            &mut self.channel.rx
+        }
+
+        fn chain(&mut self) -> &mut DescriptorChain {
+            &mut self.tx_chain
```

Comment:
Sorry I know this is an old change, but is this a typo, or why are we using the tx_chain for rx support?

Reply:
unintentional I think, but.... how does this even work?

Reply:
that method must not actually be used during the xfers.